File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Cdm/CdmDataPartitionPatternDefinition.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Cdm
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text.RegularExpressions;
    using System.Threading.Tasks;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Storage;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Microsoft.CommonDataModel.ObjectModel.Utilities.Logging;

    /// <summary>
    /// The object model implementation for Data Partition Pattern.
    /// </summary>
    public class CdmDataPartitionPatternDefinition : CdmObjectDefinitionBase, CdmFileStatus
    {
        private static readonly string Tag = nameof(CdmDataPartitionPatternDefinition);

        /// <summary>
        /// Regex timeout value
        /// </summary>
        private static TimeSpan DefaultRegexTimeoutValue = TimeSpan.FromSeconds(1);

        private TraitToPropertyMap TraitToPropertyMap { get; }

        /// <summary>
        /// Initializes a new instance of the <see cref="CdmDataPartitionPatternDefinition"/> class.
        /// </summary>
        /// <param name="ctx">The context.</param>
        /// <param name="name">The name.</param>
        public CdmDataPartitionPatternDefinition(CdmCorpusContext ctx, string name) : base(ctx)
        {
            this.ObjectType = CdmObjectType.DataPartitionPatternDef;
            this.Name = name;
            this.TraitToPropertyMap = new TraitToPropertyMap(this);
        }

        /// <summary>
        /// Gets or sets the name of the data partition pattern.
        /// </summary>
        public string Name { get; set; }

        /// <summary>
        /// Gets or sets the starting location corpus path to use to search for inferred data partitions.
        /// </summary>
        public string RootLocation { get; set; }

        /// <summary>
        /// Gets or sets the glob pattern string to use to search for partitions.
        /// If both globPattern and regularExpression is set, globPattern will be used.
        /// </summary>
        public string GlobPattern { get; set; }

        /// <summary>
        /// Gets or sets the regular expression string to use to search for partitions.
        /// </summary>
        public string RegularExpression { get; set; }

        /// <summary>
        /// Gets or sets the names for replacement values from the regular expression.
        /// </summary>
        public List<string> Parameters { get; set; }

        /// <summary>
        /// Gets or sets the corpus path for the specialized schema to use for matched pattern partitions.
        /// </summary>
        public string SpecializedSchema { get; set; }

        /// <summary>
        /// Gets or sets the last file status check time.
        /// </summary>
        public DateTimeOffset? LastFileStatusCheckTime { get; set; }

        /// <summary>
        /// Gets or sets the last file modified time.
        /// </summary>
        public DateTimeOffset? LastFileModifiedTime { get; set; }

        /// <summary>
        /// LastChildFileModifiedTime is not valid for DataPartitionPatterns since they do not contain any children objects.
        /// </summary>
        public DateTimeOffset? LastChildFileModifiedTime { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }

        /// <summary>
        /// Gets or sets whether the data partition pattern is incremental.
        /// </summary>
        public bool IsIncremental { get => this.TraitToPropertyMap.FetchPropertyValue(nameof(IsIncremental).Substring(0, 1).ToLower() + nameof(IsIncremental).Substring(1)); }

        /// <inheritdoc />
        [Obsolete]
        public override CdmObjectType GetObjectType()
        {
            return CdmObjectType.DataPartitionPatternDef;
        }

        /// <inheritdoc />
        public override bool Validate()
        {
            if (string.IsNullOrWhiteSpace(this.RootLocation))
            {
                IEnumerable<string> missingFields = new List<string>() { "RootLocation" };
                Logger.Error(this.Ctx, Tag, nameof(Validate), this.AtCorpusPath, CdmLogCode.ErrValdnIntegrityCheckFailure, this.AtCorpusPath, string.Join(", ", missingFields.Select((s) => $"'{s}'")));
                return false;
            }
            return true;
        }

        /// <inheritdoc />
        public override CdmObject Copy(ResolveOptions resOpt = null, CdmObject host = null)
        {
            if (resOpt == null)
            {
                resOpt = new ResolveOptions(this, this.Ctx.Corpus.DefaultResolutionDirectives);
            }

            CdmDataPartitionPatternDefinition copy;
            if (host == null)
            {
                copy = new CdmDataPartitionPatternDefinition(this.Ctx, this.Name);
            }
            else
            {
                copy = host as CdmDataPartitionPatternDefinition;
                copy.Name = this.Name;
            }

            copy.RootLocation = this.RootLocation;
            copy.GlobPattern = this.GlobPattern;
            copy.RegularExpression = this.RegularExpression;
            copy.Parameters = this.Parameters != null ? new List<String>(this.Parameters) : null;
            copy.LastFileStatusCheckTime = this.LastFileStatusCheckTime;
            copy.LastFileModifiedTime = this.LastFileModifiedTime;

            if (this.SpecializedSchema != null)
            {
                copy.SpecializedSchema = this.SpecializedSchema;
            }

            this.CopyDef(resOpt, copy);

            return copy;
        }

        /// <inheritdoc />
        public override string GetName()
        {
            return this.Name;
        }

        [Obsolete("CopyData is deprecated. Please use the Persistence Layer instead.")]
        public override dynamic CopyData(ResolveOptions resOpt, CopyOptions options)
        {
            return CdmObjectBase.CopyData<CdmDataPartitionPatternDefinition>(this, resOpt, options);
        }

        /// <inheritdoc />
        public override bool Visit(string pathFrom, VisitCallback preChildren, VisitCallback postChildren)
        {
            string path = this.UpdateDeclaredPath(pathFrom);

            if (preChildren != null && preChildren.Invoke(this, path))
                return false;

            if (this.VisitDef(path, preChildren, postChildren))
                return true;

            if (postChildren != null && postChildren.Invoke(this, path))
                return true;
            return false;
        }

        /// <inheritdoc />
        internal override string UpdateDeclaredPath(string pathFrom)
        {
            return pathFrom + this.GetName() ?? "UNNAMED";
        }

        /// <inheritdoc />
        public override bool IsDerivedFrom(string baseName, ResolveOptions resOpt)
        {
            return false;
        }

        /// <inheritdoc />
        public async Task FileStatusCheckAsync(FileStatusCheckOptions fileStatusCheckOptions)
        {
            await this.FileStatusCheckAsyncInternal(fileStatusCheckOptions);
        }

        internal async Task<bool> FileStatusCheckAsyncInternal(FileStatusCheckOptions fileStatusCheckOptions)
        {
            using (Logger.EnterScope(nameof(CdmDataPartitionPatternDefinition), Ctx, nameof(FileStatusCheckAsync)))
            {
                string nameSpace = null;
                StorageAdapterBase adapter = null;

                // make sure the root is a good full corpus path
                string rootCleaned = this.RootLocation?.EndsWith("/") == true ? this.RootLocation.Substring(0, this.RootLocation.Length - 1) : this.RootLocation;
                if (rootCleaned == null)
                {
                    rootCleaned = "";
                }
                string rootCorpus = this.Ctx.Corpus.Storage.CreateAbsoluteCorpusPath(rootCleaned, this.InDocument);

                IDictionary<string, CdmFileMetadata> fileInfoList = null;
                try
                {
                    // Remove namespace from path
                    Tuple<string, string> pathTuple = StorageUtils.SplitNamespacePath(rootCorpus);
                    if (pathTuple == null)
                    {
                        Logger.Error(this.Ctx, Tag, nameof(FileStatusCheckAsync), this.AtCorpusPath, CdmLogCode.ErrStorageNullCorpusPath);
                        return true;
                    }

                    nameSpace = pathTuple.Item1;
                    adapter = this.Ctx.Corpus.Storage.FetchAdapter(nameSpace);

                    if (adapter == null)
                    {
                        Logger.Error(this.Ctx, Tag, nameof(FileStatusCheckAsync), this.AtCorpusPath, CdmLogCode.ErrDocAdapterNotFound, this.InDocument.Name);
                        return true;
                    }

                    // get a list of all corpusPaths under the root
                    fileInfoList = await adapter.FetchAllFilesMetadataAsync(pathTuple.Item2);
                }
                catch (Exception e)
                {
                    Logger.Warning(this.Ctx, Tag, nameof(FileStatusCheckAsync), this.AtCorpusPath, CdmLogCode.WarnPartitionFileFetchFailed, rootCorpus, e.Message);
                }

                // update modified times
                this.LastFileStatusCheckTime = DateTimeOffset.UtcNow;

                if (fileInfoList == null)
                {
                    Logger.Error(this.Ctx, Tag, nameof(FileStatusCheckAsync), this.AtCorpusPath, CdmLogCode.ErrFetchingFileMetadataNull, nameSpace);
                    return true;
                }

                if (nameSpace != null)
                {
                    // remove root of the search from the beginning of all paths so anything in the root is not found by regex
                    IDictionary<string, CdmFileMetadata> cleanedFileList = new Dictionary<string, CdmFileMetadata>();

                    foreach (var entry in fileInfoList)
                    {
                        string newFileName = $"{nameSpace}:{entry.Key}";
                        newFileName = newFileName.Slice(rootCorpus.Length);
                        cleanedFileList.Add(newFileName, entry.Value);
                    }

                    if (this.Owner is CdmLocalEntityDeclarationDefinition localEntDecDefOwner)
                    {
                        // if both are present log warning and use glob pattern, otherwise use regularExpression
                        if (!String.IsNullOrWhiteSpace(this.GlobPattern) && !String.IsNullOrWhiteSpace(this.RegularExpression))
                        {
                            Logger.Warning(this.Ctx,
                                Tag,
                                nameof(FileStatusCheckAsync),
                                this.AtCorpusPath,
                                CdmLogCode.WarnPartitionGlobAndRegexPresent,
                                this.GlobPattern,
                                this.RegularExpression);
                        }
                        string regularExpression = !String.IsNullOrWhiteSpace(this.GlobPattern) ? this.GlobPatternToRegex(this.GlobPattern) : this.RegularExpression;
                        Regex regexPattern = null;

                        TimeSpan regexTimeoutValue = fileStatusCheckOptions?.RegexTimeoutSeconds != null ? TimeSpan.FromSeconds(fileStatusCheckOptions.RegexTimeoutSeconds.Value) : DefaultRegexTimeoutValue;

                        try
                        {
                            regexPattern = new Regex(regularExpression, RegexOptions.None, regexTimeoutValue);
                        }
                        catch (ArgumentOutOfRangeException rangeEx)
                        {
                            Logger.Error(
                                this.Ctx,
                                Tag,
                                nameof(FileStatusCheckAsync),
                                this.AtCorpusPath,
                                CdmLogCode.ErrRegexTimeoutOutOfRange, regexTimeoutValue.ToString(), rangeEx.Message);
                        }
                        catch (Exception e)
                        {
                            Logger.Error(this.Ctx, Tag,
                                nameof(FileStatusCheckAsync),
                                this.AtCorpusPath,
                                CdmLogCode.ErrValdnInvalidExpression, !String.IsNullOrWhiteSpace(this.GlobPattern) ? "glob pattern" : "regular expression",
                                !String.IsNullOrWhiteSpace(this.GlobPattern) ? this.GlobPattern : this.RegularExpression, e.Message);
                        }

                        if (regexPattern != null)
                        {
                            // a hashset to check if the data partition exists
                            HashSet<string> dataPartitionPathHashSet = new HashSet<string>();
                            if (localEntDecDefOwner.DataPartitions != null)
                            {
                                foreach (var dataPartition in localEntDecDefOwner.DataPartitions)
                                {
                                    var fullPath = this.Ctx.Corpus.Storage.CreateAbsoluteCorpusPath(dataPartition.Location, this.InDocument);
                                    dataPartitionPathHashSet.Add(fullPath);
                                }
                            }

                            HashSet<string> incrementalPartitionPathHashSet = new HashSet<string>();
                            if (localEntDecDefOwner.IncrementalPartitions != null)
                            {
                                foreach (var incrementalPartition in localEntDecDefOwner.IncrementalPartitions)
                                {
                                    var fullPath = this.Ctx.Corpus.Storage.CreateAbsoluteCorpusPath(incrementalPartition.Location, this.InDocument);
                                    incrementalPartitionPathHashSet.Add(fullPath);
                                }
                            }

                            foreach (var fi in cleanedFileList)
                            {
                                string fileName = fi.Key;
                                CdmFileMetadata partitionMetadata = fi.Value;

                                Match m;

                                try
                                {
                                    m = regexPattern.Match(fileName);
                                }
                                catch (RegexMatchTimeoutException e)
                                {
                                    Logger.Error(this.Ctx, Tag, nameof(FileStatusCheckAsync), this.AtCorpusPath, CdmLogCode.ErrRegexTimeout, e.Message);

                                    // do not continue processing the manifest/entity/partition pattern if timeout
                                    return false;
                                }

                                if (m.Success && m.Length > 1 && m.Value == fileName)
                                {
                                    // create a map of arguments out of capture groups
                                    Dictionary<string, List<string>> args = new Dictionary<string, List<string>>();
                                    int iParam = 0;
                                    // captures start after the string match at m.Groups[0]
                                    for (int i = 1; i < m.Groups.Count; i++)
                                    {
                                        CaptureCollection captures = m.Groups[i].Captures;
                                        if (captures.Count > 0 && iParam < this.Parameters?.Count)
                                        {
                                            // to be consistent with other languages, if a capture group captures
                                            // multiple things, only use the last thing that was captured
                                            var singleCapture = captures[captures.Count - 1];

                                            string currentParam = this.Parameters[iParam];
                                            if (!args.ContainsKey(currentParam))
                                                args[currentParam] = new List<string>();
                                            args[currentParam].Add(singleCapture.ToString());
                                            iParam++;
                                        }
                                        else
                                        {
                                            break;
                                        }
                                    }
                                    // put the original but cleaned up root back onto the matched doc as the location stored in the partition
                                    string locationCorpusPath = $"{rootCleaned}{fileName}";
                                    string fullPath = $"{rootCorpus}{fileName}";
                                    // Remove namespace from path
                                    Tuple<string, string> pathTuple = StorageUtils.SplitNamespacePath(fullPath);
                                    if (pathTuple == null)
                                    {
                                        Logger.Error(this.Ctx, Tag, nameof(FileStatusCheckAsync), this.AtCorpusPath, CdmLogCode.ErrStorageNullCorpusPath, this.AtCorpusPath);
                                        return true;
                                    }

                                    CdmTraitCollection exhibitsTraits = this.ExhibitsTraits;
                                    if (fileStatusCheckOptions?.IncludeDataPartitionSize == true && partitionMetadata?.FileSizeBytes != null)
                                    {
                                        exhibitsTraits = new CdmTraitCollection(this.Ctx, this);
                                        foreach (var trait in this.ExhibitsTraits)
                                        {
                                            exhibitsTraits.Add(trait);
                                        }
                                        exhibitsTraits.Add("is.partition.size", new List<Tuple<string, dynamic>> { new Tuple<string, dynamic>("value", partitionMetadata.FileSizeBytes) });
                                    }

                                    try
                                    {
                                        DateTimeOffset? lastModifiedTime = await adapter.ComputeLastModifiedTimeAsync(pathTuple.Item2);
                                        if (this.IsIncremental && !incrementalPartitionPathHashSet.Contains(fullPath))
                                        {
                                            localEntDecDefOwner.CreateDataPartitionFromPattern(locationCorpusPath, exhibitsTraits, args, this.SpecializedSchema, lastModifiedTime, true, this.Name);
                                            incrementalPartitionPathHashSet.Add(fullPath);
                                        }
                                        else if (!this.IsIncremental && !dataPartitionPathHashSet.Contains(fullPath))
                                        {
                                            localEntDecDefOwner.CreateDataPartitionFromPattern(locationCorpusPath, exhibitsTraits, args, this.SpecializedSchema, lastModifiedTime);
                                            dataPartitionPathHashSet.Add(fullPath);
                                        }
                                    }
                                    catch (Exception e)
                                    {
                                        Logger.Error(this.Ctx, Tag, nameof(FileStatusCheckAsync), this.AtCorpusPath, CdmLogCode.ErrPartitionFileModTimeFailure, pathTuple.Item2, e.Message);
                                    }
                                }
                            }
                        }
                    }
                }
                return true;
            }
        }

        /// <inheritdoc />
        public async Task FileStatusCheckAsync()
        {
            await this.FileStatusCheckAsync(null);
        }

        /// <inheritdoc />
        public async Task ReportMostRecentTimeAsync(DateTimeOffset? childTime)
        {
            if (this.Owner is CdmFileStatus && childTime != null)
                await (this.Owner as CdmFileStatus).ReportMostRecentTimeAsync(childTime);
        }

        /// <summary>
        /// Converts a glob pattern to a regular expression 
        /// </summary>
        private string GlobPatternToRegex(string pattern)
        {
            List<string> newPattern = new List<string>();

            // all patterns should start with a slash
            newPattern.Add("[/\\\\]");

            // if pattern starts with slash, skip the first character. We already added it above
            for (int i = (pattern[0] == '/' || pattern[0] == '\\' ? 1 : 0); i < pattern.Length; i++)
            {
                char currChar = pattern[i];

                switch (currChar)
                {
                    case '.':
                        // escape '.' characters
                        newPattern.Add("\\.");
                        break;
                    case '\\':
                        // convert backslash into slash
                        newPattern.Add("[/\\\\]");
                        break;
                    case '?':
                        // question mark in glob matches any single character
                        newPattern.Add(".");
                        break;
                    case '*':
                        char? nextChar = i + 1 < pattern.Length ? pattern[i + 1] : (char?)null;
                        if (nextChar != null && nextChar == '*')
                        {
                            char? prevChar = i - 1 >= 0 ? pattern[i - 1] : (char?)null;
                            char? postChar = i + 2 < pattern.Length ? pattern[i + 2] : (char?)null;

                            // globstar must be at beginning of pattern, end of pattern, or wrapped in separator characters
                            if ((prevChar == null || prevChar == '/' || prevChar == '\\')
                            && (postChar == null || postChar == '/' || postChar == '\\'))
                            {
                                newPattern.Add(".*");

                                // globstar can match zero or more subdirectories. If it matches zero, then there should not be
                                // two consecutive '/' characters so make the second one optional
                                if (prevChar != null && postChar != null &&
                                (prevChar == '/' || prevChar == '\\') && (postChar == '/' || postChar == '\\'))
                                {
                                    newPattern.Add("/?");
                                    i++;
                                }
                            }
                            else
                            {
                                // otherwise, treat the same as '*'
                                newPattern.Add("[^/\\\\]*");
                            }
                            i++;
                        }
                        else
                        {
                            // *
                            newPattern.Add("[^/\\\\]*");
                        }
                        break;
                    default:
                        newPattern.Add(currChar.ToString());
                        break;
                }
            }

            return string.Join("", newPattern);
        }
    }
}

File: objectModel/TypeScript/Cdm/CdmDataPartitionPatternDefinition.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import {
    CdmCorpusContext,
    CdmFileMetadata,
    CdmFileStatus,
    CdmLocalEntityDeclarationDefinition,
    cdmLogCode,
    CdmObject,
    CdmObjectDefinitionBase,
    cdmObjectType,
    CdmTraitCollection,
    fileStatusCheckOptions,
    resolveOptions,
    StorageAdapterBase,
    traitToPropertyMap,
    VisitCallback
} from '../internal';
import { isLocalEntityDeclarationDefinition } from '../Utilities/cdmObjectTypeGuards';
import { Logger, enterScope } from '../Utilities/Logging/Logger';
import { StorageUtils } from '../Utilities/StorageUtils';
import { using } from "using-statement";
import path = require('node:path');

/**
 * The object model implementation for Data Partition Pattern.
 */
export class CdmDataPartitionPatternDefinition extends CdmObjectDefinitionBase implements CdmFileStatus {
    private TAG: string = CdmDataPartitionPatternDefinition.name;

    /**
     * The name of the data partition pattern.
     */
    public name: string;

    /**
     * Gets or sets the starting location corpus path to use to search for inferred data partitions.
     */
    public rootLocation: string;

    /**
     * Gets or sets the glob pattern used to search for partitions.
     * If both globPattern and regularExpression is set, globPattern will be used.
     */
    public globPattern?: string;

    /**
     * Gets or sets the regular expression string to use to search for partitions.
     */
    public regularExpression?: string;

    /**
     * Gets or sets the names for replacement values from the regular expression.
     */
    public parameters?: string[];

    /**
     * Gets or sets the corpus path for the specialized schema to use for matched pattern partitions.
     */
    public specializedSchema?: string;

    /**
     * @inheritdoc
     */
    public lastFileStatusCheckTime: Date;

    /**
     * @inheritdoc
     */
    public lastFileModifiedTime: Date;

    /**
     * LastChildFileModifiedTime is not valid for DataPartitionPatterns since they do not contain any children objects.
     */
    public get lastChildFileModifiedTime(): Date {
        throw new Error('Not implemented');
    }

    /**
     * LastChildFileModifiedTime is not valid for DataPartitionPatterns since they do not contain any children objects.
     */
    public set lastChildFileModifiedTime(time: Date) {
        throw new Error('Not implemented');
    }

    public static get objectType(): cdmObjectType {
        return cdmObjectType.dataPartitionPatternDef;
    }

    private readonly traitToPropertyMap: traitToPropertyMap;

    /**
     * Creates a new instance of Data Partition Pattern Impl.
     * @param ctx The context.
     * @param name The name.
     */
    constructor(ctx: CdmCorpusContext, name: string) {
        super(ctx);
        this.objectType = cdmObjectType.dataPartitionPatternDef;
        this.name = name;
        this.traitToPropertyMap = new traitToPropertyMap(this);
    }

    /**
     * Gets whether the data partition pattern is incremental.
     */
    public get isIncremental(): boolean {
        return this.traitToPropertyMap.fetchPropertyValue('isIncremental') as boolean;
    }

    /**
     * @inheritdoc
     */
    public getObjectType(): cdmObjectType {
        return cdmObjectType.dataPartitionPatternDef;
    }

    /**
     * @inheritdoc
     */
    public validate(): boolean {
        if (!this.rootLocation) {
            let missingFields: string[] = ['rootLocation'];
            Logger.error(this.ctx, this.TAG, this.validate.name, this.atCorpusPath, cdmLogCode.ErrValdnIntegrityCheckFailure, missingFields.map((s: string) => `'${s}'`).join(', '), this.atCorpusPath);
            return false;
        }

        return true;
    }

    /**
     * @inheritdoc
     */
    public copy(resOpt: resolveOptions, host?: CdmObject): CdmDataPartitionPatternDefinition {
        if (resOpt === undefined) {
            resOpt = new resolveOptions(this, this.ctx.corpus.defaultResolutionDirectives);
        }
        let copy: CdmDataPartitionPatternDefinition;
        if (!host) {
            copy = new CdmDataPartitionPatternDefinition(this.ctx, this.name);
        } else {
            copy = host as CdmDataPartitionPatternDefinition;
            copy.name = this.name;
        }
        copy.rootLocation = this.rootLocation;
        copy.globPattern = this.globPattern;
        copy.regularExpression = this.regularExpression;
        copy.lastFileStatusCheckTime = this.lastFileStatusCheckTime;
        copy.lastFileModifiedTime = this.lastFileModifiedTime;
        copy.parameters = this.parameters ? this.parameters.slice() : undefined;
        if (this.specializedSchema) {
            copy.specializedSchema = this.specializedSchema;
        }
        this.copyDef(resOpt, copy);

        return copy;
    }

    /**
     * @inheritdoc
     */
    public getName(): string {
        return this.name;
    }

    /**
     * @inheritdoc
     */
    public visit(pathFrom: string, preChildren: VisitCallback, postChildren: VisitCallback): boolean {
        const path: string = this.fetchDeclaredPath(pathFrom);

        if (preChildren && preChildren(this, path)) {
            return false;
        }

        if (this.visitDef(path, preChildren, postChildren)) {
            return true;
        }

        if (postChildren && postChildren(this, path)) {
            return true;
        }

        return false;
    }

    /**
     * @internal
     */
    public fetchDeclaredPath(pathFrom: string): string {
        return pathFrom + (this.getName() || 'UNNAMED');
    }

    /**
     * @inheritdoc
     */
    public isDerivedFrom(base: string, resOpt?: resolveOptions): boolean {
        return false;
    }

    /**
     * @inheritdoc
     */
    public async fileStatusCheckAsync(fileStatusCheckOptions?: fileStatusCheckOptions): Promise<void> {
        return await using(enterScope(CdmDataPartitionPatternDefinition.name, this.ctx, this.fileStatusCheckAsync.name), async _ => {
            let namespace: string = undefined;
            let adapter: StorageAdapterBase = undefined;

            // make sure the root is a good full corpus path
            let rootCleaned: string = this.rootLocation && this.rootLocation.endsWith('/') ? this.rootLocation.substring(0, this.rootLocation.length - 1) : this.rootLocation;
            if (rootCleaned === undefined) {
                rootCleaned = '';
            }
            const rootCorpus: string = this.ctx.corpus.storage.createAbsoluteCorpusPath(rootCleaned, this.inDocument);

            let fileInfoList: Map<string, CdmFileMetadata>;
            try {
                // Remove namespace from path
                const pathTuple: [string, string] = StorageUtils.splitNamespacePath(rootCorpus);
                if (!pathTuple) {
                    Logger.error(this.ctx, this.TAG, this.fileStatusCheckAsync.name, this.atCorpusPath, cdmLogCode.ErrStorageNullCorpusPath);
                    return;
                }

                namespace = pathTuple[0];
                adapter = this.ctx.corpus.storage.fetchAdapter(namespace);

                if (adapter === undefined) {
                    Logger.error(this.ctx, this.TAG, this.fileStatusCheckAsync.name, this.atCorpusPath, cdmLogCode.ErrDocAdapterNotFound, this.inDocument.name);
                    return;
                }

                // get a list of all corpusPaths under the root
                fileInfoList = await adapter.fetchAllFilesMetadataAsync(pathTuple[1]);
            } catch (e) {
                Logger.warning(this.ctx, this.TAG, this.fileStatusCheckAsync.name, this.atCorpusPath, cdmLogCode.WarnPartitionFileFetchFailed, rootCorpus, e.Message);
            }

            // update modified times
            this.lastFileStatusCheckTime = new Date();

            if (fileInfoList === undefined) {
                Logger.error(this.ctx, this.TAG, this.fileStatusCheckAsync.name, this.atCorpusPath, cdmLogCode.ErrFetchingFileMetadataNull, namespace);
                return;
            }

            if (namespace !== undefined) {
                // remove root of the search from the beginning of all paths so anything in the root is not found by regex
                const cleanedFileList: Map<string, CdmFileMetadata> = new Map<string, CdmFileMetadata>();

                for (const entry of fileInfoList) {
                    let newFileName: string = `${namespace}:${entry[0]}`;
                    newFileName = newFileName.slice(rootCorpus.length);
                    cleanedFileList.set(newFileName, entry[1]);
                }

                if (isLocalEntityDeclarationDefinition(this.owner)) {
                    const localEntDecDefOwner: CdmLocalEntityDeclarationDefinition = this.owner;
                    // if both are present log warning and use glob pattern, otherwise use regularExpression
                    if (this.globPattern && this.globPattern.trim() !== '' && this.regularExpression && this.regularExpression.trim() !== '') {
                        Logger.warning(this.ctx, this.TAG, this.fileStatusCheckAsync.name, this.atCorpusPath, cdmLogCode.WarnPartitionGlobAndRegexPresent, this.globPattern, this.regularExpression);
                    }
                    const regularExpression: string =
                        this.globPattern && this.globPattern.trim() !== '' ? this.globPatternToRegex(this.globPattern) : this.regularExpression;
                    let regexPattern: RegExp;

                    try {
                        regexPattern = new RegExp(regularExpression);
                    } catch (e) {
                        Logger.error(this.ctx, this.TAG, this.fileStatusCheckAsync.name, this.atCorpusPath, cdmLogCode.ErrValdnInvalidExpression, this.globPattern, this.regularExpression, e.message);
                    }

                    if (regexPattern !== undefined) {
                        const dataPartitionPathSet: Set<string> = new Set<string>();
                        if (localEntDecDefOwner.dataPartitions) {
                            for (const dataPartition of localEntDecDefOwner.dataPartitions) {
                                const fullPath: string = this.ctx.corpus.storage.createAbsoluteCorpusPath(dataPartition.location, this.inDocument);
                                dataPartitionPathSet.add(fullPath);
                            }
                        }

                        const incrementalPartitionPathHashSet: Set<string> = new Set<string>();
                        if (localEntDecDefOwner.incrementalPartitions) {
                            for (const incrementalPartition of localEntDecDefOwner.incrementalPartitions) {
                                const fullPath: string = this.ctx.corpus.storage.createAbsoluteCorpusPath(incrementalPartition.location, this.inDocument);
                                incrementalPartitionPathHashSet.add(fullPath);
                            }
                        }

                        for (const fi of cleanedFileList) {
                            const fileName: string = fi[0];
                            const partitionMetadata: CdmFileMetadata = fi[1];

                            const m: RegExpExecArray = regexPattern.exec(fileName);
                            if (m && m.length > 0 && m[0] === fileName) {
                                // create a map of arguments out of capture groups
                                const args: Map<string, string[]> = new Map();
                                // captures start after the string match at m[0]
                                for (let i: number = 1; i < m.length; i++) {
                                    const iParam: number = i - 1;
                                    if (this.parameters && iParam < this.parameters.length) {
                                        const currentParam: string = this.parameters[iParam];
                                        if (!args.has(currentParam)) {
                                            args.set(currentParam, []);
                                        }
                                        args.get(currentParam)
                                            .push(m[i]);
                                    }
                                }

                                // put the origial but cleaned up root back onto the matched doc as the location stored in the partition
                                const locationCorpusPath: string = `${rootCleaned}${fileName}`;
                                const fullPath: string = `${rootCorpus}${fileName}`;
                                // Remove namespace from path
                                const pathTuple: [string, string] = StorageUtils.splitNamespacePath(fullPath);
                                if (!pathTuple) {
                                    Logger.error(this.ctx, this.TAG, this.fileStatusCheckAsync.name, this.atCorpusPath, cdmLogCode.ErrStorageNullCorpusPath);
                                    return;
                                }

                                let exhibitsTraits: CdmTraitCollection = this.exhibitsTraits;
                                if (fileStatusCheckOptions?.includeDataPartitionSize && partitionMetadata?.fileSizeBytes != undefined) {
                                    exhibitsTraits = new CdmTraitCollection(this.ctx, this);
                                    for (const trait of this.exhibitsTraits) {
                                        exhibitsTraits.push(trait);
                                    }

                                    exhibitsTraits.push('is.partition.size', [['value', partitionMetadata.fileSizeBytes]]);
                                }

                                const lastModifiedTime: Date = await adapter.computeLastModifiedTimeAsync(pathTuple[1]);

                                if (this.isIncremental && !incrementalPartitionPathHashSet.has(fullPath)) {
                                    localEntDecDefOwner.createDataPartitionFromPattern(
                                        locationCorpusPath, exhibitsTraits, args, this.specializedSchema, lastModifiedTime, true, this.name);
                                    dataPartitionPathSet.add(fullPath);
                                } else if (!this.isIncremental && !dataPartitionPathSet.has(fullPath)) {
                                    localEntDecDefOwner.createDataPartitionFromPattern(
                                        locationCorpusPath, exhibitsTraits, args, this.specializedSchema, lastModifiedTime);
                                    dataPartitionPathSet.add(fullPath);
                                }
                            }
                        }
                    }
                }
            }
        });
    }

    /**
     * @inheritdoc
     */
    public async reportMostRecentTimeAsync(childTime: Date): Promise<void> {
        if ((this.owner as CdmFileStatus).reportMostRecentTimeAsync && childTime) {
            await (this.owner as CdmFileStatus).reportMostRecentTimeAsync(childTime);
        }
    }

    /**
     * Converts a glob pattern to a regular expression
     */
    private globPatternToRegex(pattern: string): string {
        const newPattern: string[] = [];

        // all patterns should start with a slash
        newPattern.push("[/\\\\]");

        // if pattern starts with slash, skip the first character. We already added it above
        for (let i: number = (pattern[0] === '/' || pattern[0] === '\\' ? 1 : 0); i < pattern.length; i++) {
            const currChar: string = pattern[i];

            switch (currChar) {
                case '.':
                    // escape '.' characters
                    newPattern.push('\\.');
                    break;
                case '\\':
                    // convert backslash into slash
                    newPattern.push('[/\\\\]');
                    break;
                case '?':
                    // question mark in glob matches any single character
                    newPattern.push('.');
                    break;
                case '*':
                    const nextChar: string = i + 1 < pattern.length ? pattern[i + 1] : undefined;
                    if (nextChar === '*') {
                        const prevChar: string = i - 1 >= 0 ? pattern[i - 1] : undefined;
                        const postChar: string = i + 2 < pattern.length ? pattern[i + 2] : undefined;

                        // globstar must be at beginning of pattern, end of pattern, or wrapped in separator characters
                        if ((prevChar === undefined || prevChar === '/' || prevChar === '\\')
                            && (postChar === undefined || postChar === '/' || postChar === '\\')) {
                            newPattern.push('.*');

                            // globstar can match zero or more subdirectories. If it matches zero, then there should not be
                            // two consecutive '/' characters so make the second one optional
                            if ((prevChar === '/' || prevChar === '\\') && (postChar === '/' || postChar === '\\')) {
                                newPattern.push('/?');
                                i++;
                            }
                        } else {
                            // otherwise, treat the same as '*'
                            newPattern.push('[^\/\\\\]*');
                        }
                        i++;
                    } else {
                        // *
                        newPattern.push('[^\/\\\\]*');
                    }
                    break;
                default:
                    newPattern.push(currChar);
            }
        }

        return newPattern.join('');
    }
}
